File: app/src/main/java/org/wikipedia/theme/ThemeChooserDialog.java
package org.wikipedia.theme;

import android.content.DialogInterface;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SwitchCompat;
import androidx.core.content.ContextCompat;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.button.MaterialButton;

import org.wikipedia.Constants;
import org.wikipedia.R;
import org.wikipedia.WikipediaApp;
import org.wikipedia.activity.FragmentUtil;
import org.wikipedia.analytics.AppearanceChangeFunnel;
import org.wikipedia.events.WebViewInvalidateEvent;
import org.wikipedia.page.ExtendedBottomSheetDialogFragment;
import org.wikipedia.settings.Prefs;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.FeedbackUtil;
import org.wikipedia.util.ResourceUtil;
import org.wikipedia.views.DiscreteSeekBar;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.Unbinder;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.functions.Consumer;

import static org.wikipedia.Constants.INTENT_EXTRA_INVOKE_SOURCE;

public class ThemeChooserDialog extends ExtendedBottomSheetDialogFragment {
    @BindView(R.id.buttonDecreaseTextSize) TextView buttonDecreaseTextSize;
    @BindView(R.id.buttonIncreaseTextSize) TextView buttonIncreaseTextSize;
    @BindView(R.id.text_size_percent) TextView textSizePercent;
    @BindView(R.id.text_size_seek_bar) DiscreteSeekBar textSizeSeekBar;
    @BindView(R.id.button_font_family_sans_serif) MaterialButton buttonFontFamilySansSerif;
    @BindView(R.id.button_font_family_serif) MaterialButton buttonFontFamilySerif;
    @BindView(R.id.button_theme_light) MaterialButton buttonThemeLight;
    @BindView(R.id.button_theme_dark) MaterialButton buttonThemeDark;
    @BindView(R.id.button_theme_black) MaterialButton buttonThemeBlack;
    @BindView(R.id.button_theme_sepia) MaterialButton buttonThemeSepia;
    @BindView(R.id.theme_chooser_dark_mode_dim_images_switch) SwitchCompat dimImagesSwitch;
    @BindView(R.id.theme_chooser_match_system_theme_switch) SwitchCompat matchSystemThemeSwitch;
    @BindView(R.id.font_change_progress_bar) ProgressBar fontChangeProgressBar;

    public interface Callback {
        void onToggleDimImages();
        void onCancelThemeChooser();
    }

    private enum FontSizeAction { INCREASE, DECREASE, RESET }

    private WikipediaApp app;
    private Unbinder unbinder;
    private AppearanceChangeFunnel funnel;
    private Constants.InvokeSource invokeSource;
    private CompositeDisposable disposables = new CompositeDisposable();
    private static final int BUTTON_STROKE_WIDTH = DimenUtil.roundedDpToPx(2f);

    private boolean updatingFont = false;

    public static ThemeChooserDialog newInstance(@NonNull Constants.InvokeSource source) {
        ThemeChooserDialog dialog = new ThemeChooserDialog();
        Bundle args = new Bundle();
        args.putSerializable(INTENT_EXTRA_INVOKE_SOURCE, source);
        dialog.setArguments(args);
        return dialog;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.dialog_theme_chooser, container);
        unbinder = ButterKnife.bind(this, rootView);
        buttonDecreaseTextSize.setOnClickListener(new FontSizeButtonListener(FontSizeAction.DECREASE));
        buttonIncreaseTextSize.setOnClickListener(new FontSizeButtonListener(FontSizeAction.INCREASE));
        FeedbackUtil.setButtonLongPressToast(buttonDecreaseTextSize, buttonIncreaseTextSize);
        buttonThemeLight.setOnClickListener(new ThemeButtonListener(Theme.LIGHT));
        buttonThemeDark.setOnClickListener(new ThemeButtonListener(Theme.DARK));
        buttonThemeBlack.setOnClickListener(new ThemeButtonListener(Theme.BLACK));
        buttonThemeSepia.setOnClickListener(new ThemeButtonListener(Theme.SEPIA));
        buttonFontFamilySansSerif.setOnClickListener(new FontFamilyListener());
        buttonFontFamilySerif.setOnClickListener(new FontFamilyListener());

        textSizeSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int value, boolean fromUser) {
                if (!fromUser) {
                    return;
                }
                int currentMultiplier = Prefs.getTextSizeMultiplier();
                boolean changed = app.setFontSizeMultiplier(textSizeSeekBar.getValue());
                if (changed) {
                    updatingFont = true;
                    updateFontSize();
                    funnel.logFontSizeChange(currentMultiplier, Prefs.getTextSizeMultiplier());
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        updateComponents();
        disableBackgroundDim();
        setNavigationBarColor(ResourceUtil.getThemedColor(requireContext(), R.attr.paper_color));
        return rootView;
    }

    @Override
    public void onStart() {
        super.onStart();
        BottomSheetBehavior.from((View) getView().getParent()).setPeekHeight(DimenUtil
                .roundedDpToPx(DimenUtil.getDimension(R.dimen.themeChooserSheetPeekHeight)));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        app = WikipediaApp.getInstance();
        invokeSource = (Constants.InvokeSource) getArguments().getSerializable(INTENT_EXTRA_INVOKE_SOURCE);
        disposables.add(app.getBus().subscribe(new EventBusConsumer()));
        funnel = new AppearanceChangeFunnel(app, app.getWikiSite(), invokeSource);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disposables.clear();
    }

    @Override
    public void onCancel(@NonNull DialogInterface dialog) {
        super.onCancel(dialog);
        if (callback() != null) {
            // noinspection ConstantConditions
            callback().onCancelThemeChooser();
        }
    }

    @OnCheckedChanged(R.id.theme_chooser_dark_mode_dim_images_switch)
    void onToggleDimImages(boolean enabled) {
        if (enabled == Prefs.shouldDimDarkModeImages()) {
            return;
        }
        Prefs.setDimDarkModeImages(enabled);
        if (callback() != null) {
            // noinspection ConstantConditions
            callback().onToggleDimImages();
        }
    }

    @OnCheckedChanged(R.id.theme_chooser_match_system_theme_switch)
    void onToggleMatchSystemTheme(boolean enabled) {
        if (enabled == Prefs.shouldMatchSystemTheme()) {
            return;
        }
        Prefs.setMatchSystemTheme(enabled);
        Theme currentTheme = app.getCurrentTheme();
        if (isMatchingSystemThemeEnabled()) {
            switch (WikipediaApp.getInstance().getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK) {
                case Configuration.UI_MODE_NIGHT_YES:
                    if (!WikipediaApp.getInstance().getCurrentTheme().isDark()) {
                        app.setCurrentTheme(!app.unmarshalTheme(Prefs.getPreviousThemeId()).isDark() ? Theme.BLACK : app.unmarshalTheme(Prefs.getPreviousThemeId()));
                        Prefs.setPreviousThemeId(currentTheme.getMarshallingId());
                    }
                    break;
                case Configuration.UI_MODE_NIGHT_NO:
                    if (WikipediaApp.getInstance().getCurrentTheme().isDark()) {
                        app.setCurrentTheme(app.unmarshalTheme(Prefs.getPreviousThemeId()).isDark() ? Theme.LIGHT : app.unmarshalTheme(Prefs.getPreviousThemeId()));
                        Prefs.setPreviousThemeId(currentTheme.getMarshallingId());
                    }
                    break;
                default:
                    break;
            }
        }
        conditionallyDisableThemeButtons();
    }

    @SuppressWarnings("checkstyle:magicnumber")
    private void conditionallyDisableThemeButtons() {
        buttonThemeLight.setAlpha((isMatchingSystemThemeEnabled() && (WikipediaApp.getInstance().getCurrentTheme().isDark())) ? 0.2f : 1.0f);
        buttonThemeSepia.setAlpha((isMatchingSystemThemeEnabled() && (WikipediaApp.getInstance().getCurrentTheme().isDark())) ? 0.2f : 1.0f);
        buttonThemeDark.setAlpha((isMatchingSystemThemeEnabled() && (!WikipediaApp.getInstance().getCurrentTheme().isDark())) ? 0.2f : 1.0f);
        buttonThemeBlack.setAlpha((isMatchingSystemThemeEnabled() && (!WikipediaApp.getInstance().getCurrentTheme().isDark())) ? 0.2f : 1.0f);

        buttonThemeLight.setEnabled(!isMatchingSystemThemeEnabled() || (!WikipediaApp.getInstance().getCurrentTheme().isDark()));
        buttonThemeSepia.setEnabled(!isMatchingSystemThemeEnabled() || (!WikipediaApp.getInstance().getCurrentTheme().isDark()));
        buttonThemeDark.setEnabled(!isMatchingSystemThemeEnabled() || (WikipediaApp.getInstance().getCurrentTheme().isDark()));
        buttonThemeBlack.setEnabled(!isMatchingSystemThemeEnabled() || (WikipediaApp.getInstance().getCurrentTheme().isDark()));
    }

    private boolean isMatchingSystemThemeEnabled() {
        return Prefs.shouldMatchSystemTheme() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q;
    }

    private void updateComponents() {
        updateFontSize();
        updateFontFamily();
        updateThemeButtons();
        updateDimImagesSwitch();
        updateMatchSystemThemeSwitch();
    }

    private void updateMatchSystemThemeSwitch() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            matchSystemThemeSwitch.setVisibility(View.VISIBLE);
            matchSystemThemeSwitch.setChecked(Prefs.shouldMatchSystemTheme());
            conditionallyDisableThemeButtons();
        } else {
            matchSystemThemeSwitch.setVisibility(View.GONE);
        }
    }

    @SuppressWarnings("checkstyle:magicnumber")
    private void updateFontSize() {
        int mult = Prefs.getTextSizeMultiplier();
        textSizeSeekBar.setValue(mult);
        String percentStr = getString(R.string.text_size_percent,
                (int) (100 * (1 + mult * DimenUtil.getFloat(R.dimen.textSizeMultiplierFactor))));
        textSizePercent.setText(mult == 0
                ? getString(R.string.text_size_percent_default, percentStr) : percentStr);
        if (updatingFont) {
            fontChangeProgressBar.setVisibility(View.VISIBLE);
        } else {
            fontChangeProgressBar.setVisibility(View.GONE);
        }
    }

    private void updateFontFamily() {
        buttonFontFamilySansSerif.setStrokeWidth(Prefs.getFontFamily().equals(buttonFontFamilySansSerif.getTag()) ? BUTTON_STROKE_WIDTH : 0);
        buttonFontFamilySerif.setStrokeWidth(Prefs.getFontFamily().equals(buttonFontFamilySerif.getTag()) ? BUTTON_STROKE_WIDTH : 0);
    }

    private void updateThemeButtons() {
        updateThemeButtonStroke(buttonThemeLight, app.getCurrentTheme() == Theme.LIGHT);
        updateThemeButtonStroke(buttonThemeSepia, app.getCurrentTheme() == Theme.SEPIA);
        updateThemeButtonStroke(buttonThemeDark, app.getCurrentTheme() == Theme.DARK);
        updateThemeButtonStroke(buttonThemeBlack, app.getCurrentTheme() == Theme.BLACK);
    }

    private void updateThemeButtonStroke(@NonNull MaterialButton button, boolean selected) {
        button.setStrokeWidth(selected ? BUTTON_STROKE_WIDTH : 0);
        button.setClickable(!selected);
    }

    private void updateDimImagesSwitch() {
        dimImagesSwitch.setChecked(Prefs.shouldDimDarkModeImages());
        dimImagesSwitch.setEnabled(app.getCurrentTheme().isDark());
        dimImagesSwitch.setTextColor(dimImagesSwitch.isEnabled()
                ? ResourceUtil.getThemedColor(requireContext(), R.attr.section_title_color)
                : ContextCompat.getColor(requireContext(), R.color.black26));
    }

    private final class ThemeButtonListener implements View.OnClickListener {
        private Theme theme;

        private ThemeButtonListener(Theme theme) {
            this.theme = theme;
        }

        @Override
        public void onClick(View v) {
            if (app.getCurrentTheme() != theme) {
                funnel.logThemeChange(app.getCurrentTheme(), theme);
                app.setCurrentTheme(theme);
            }
        }
    }

    private final class FontFamilyListener implements View.OnClickListener {
        @Override
        public void onClick(View v) {
            if (v.getTag() != null) {
                String newFontFamily = (String) v.getTag();
                funnel.logFontThemeChange(Prefs.getFontFamily(), newFontFamily);
                app.setFontFamily(newFontFamily);
            }
        }
    }

    private final class FontSizeButtonListener implements View.OnClickListener {
        private FontSizeAction action;

        private FontSizeButtonListener(FontSizeAction action) {
            this.action = action;
        }

        @Override
        public void onClick(View view) {
            boolean changed = false;
            int currentMultiplier = Prefs.getTextSizeMultiplier();
            if (action == FontSizeAction.INCREASE) {
                changed = app.setFontSizeMultiplier(Prefs.getTextSizeMultiplier() + 1);
            } else if (action == FontSizeAction.DECREASE) {
                changed = app.setFontSizeMultiplier(Prefs.getTextSizeMultiplier() - 1);
            } else if (action == FontSizeAction.RESET) {
                changed = app.setFontSizeMultiplier(0);
            }
            if (changed) {
                updatingFont = true;
                updateFontSize();
                funnel.logFontSizeChange(currentMultiplier, Prefs.getTextSizeMultiplier());
            }
        }
    }

    private class EventBusConsumer implements Consumer<Object> {
        @Override
        public void accept(Object event) {
            if (event instanceof WebViewInvalidateEvent) {
                updatingFont = false;
                updateComponents();
            }
        }
    }

    @Nullable
    public Callback callback() {
        return FragmentUtil.getCallback(this, Callback.class);
    }
}

File: app/src/main/java/org/wikipedia/theme/ThemeChooserDialog.kt
package org.wikipedia.theme

import android.content.DialogInterface
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.core.content.ContextCompat
import androidx.core.os.bundleOf
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.button.MaterialButton
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.functions.Consumer
import org.wikipedia.Constants
import org.wikipedia.Constants.InvokeSource
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.activity.FragmentUtil
import org.wikipedia.analytics.AppearanceChangeFunnel
import org.wikipedia.databinding.DialogThemeChooserBinding
import org.wikipedia.events.WebViewInvalidateEvent
import org.wikipedia.page.ExtendedBottomSheetDialogFragment
import org.wikipedia.page.customize.CustomizeToolbarActivity
import org.wikipedia.settings.Prefs
import org.wikipedia.util.DeviceUtil
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.FeedbackUtil
import org.wikipedia.util.ResourceUtil

class ThemeChooserDialog : ExtendedBottomSheetDialogFragment() {
    private var _binding: DialogThemeChooserBinding? = null
    private val binding get() = _binding!!

    interface Callback {
        fun onToggleDimImages()
        fun onToggleReadingFocusMode()
        fun onCancelThemeChooser()
    }

    private enum class FontSizeAction {
        INCREASE, DECREASE, RESET
    }

    private var app = WikipediaApp.getInstance()
    private lateinit var funnel: AppearanceChangeFunnel
    private lateinit var invokeSource: InvokeSource
    private var isMobileWeb: Boolean = false
    private val disposables = CompositeDisposable()
    private var updatingFont = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = DialogThemeChooserBinding.inflate(inflater, container, false)
        binding.buttonDecreaseTextSize.setOnClickListener(FontSizeButtonListener(FontSizeAction.DECREASE))
        binding.buttonIncreaseTextSize.setOnClickListener(FontSizeButtonListener(FontSizeAction.INCREASE))
        FeedbackUtil.setButtonLongPressToast(binding.buttonDecreaseTextSize, binding.buttonIncreaseTextSize)
        binding.buttonThemeLight.setOnClickListener(ThemeButtonListener(Theme.LIGHT))
        binding.buttonThemeDark.setOnClickListener(ThemeButtonListener(Theme.DARK))
        binding.buttonThemeBlack.setOnClickListener(ThemeButtonListener(Theme.BLACK))
        binding.buttonThemeSepia.setOnClickListener(ThemeButtonListener(Theme.SEPIA))
        binding.buttonFontFamilySansSerif.setOnClickListener(FontFamilyListener())
        binding.buttonFontFamilySerif.setOnClickListener(FontFamilyListener())
        binding.themeChooserDarkModeDimImagesSwitch.setOnCheckedChangeListener { _, b -> onToggleDimImages(b) }
        binding.themeChooserMatchSystemThemeSwitch.setOnCheckedChangeListener { _, b -> onToggleMatchSystemTheme(b) }
        binding.themeChooserReadingFocusModeSwitch.setOnCheckedChangeListener { _, b -> onToggleReadingFocusMode(b) }
        binding.textSizeSeekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, value: Int, fromUser: Boolean) {
                if (!fromUser) {
                    return
                }
                val currentMultiplier = Prefs.textSizeMultiplier
                val changed = app.setFontSizeMultiplier(binding.textSizeSeekBar.value)
                if (changed) {
                    updatingFont = true
                    updateFontSize()
                    funnel.logFontSizeChange(currentMultiplier.toFloat(), Prefs.textSizeMultiplier.toFloat())
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        updateComponents()
        disableBackgroundDim()
        requireDialog().window?.let {
            DeviceUtil.setNavigationBarColor(it, ResourceUtil.getThemedColor(requireContext(), R.attr.paper_color))
        }

        // TODO: test only
        binding.customizeFavorites.setOnClickListener {
            startActivity(CustomizeToolbarActivity.newIntent(requireContext()))
        }

        disposables.add(WikipediaApp.getInstance().bus.subscribe(EventBusConsumer()))
        return binding.root
    }

    override fun onStart() {
        super.onStart()
        BottomSheetBehavior.from(requireView().parent as View).peekHeight = DimenUtil.roundedDpToPx(DimenUtil.getDimension(R.dimen.themeChooserSheetPeekHeight))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        invokeSource = requireArguments().getSerializable(Constants.INTENT_EXTRA_INVOKE_SOURCE) as InvokeSource
        isMobileWeb = requireArguments().getBoolean(EXTRA_IS_MOBILE_WEB)
        funnel = AppearanceChangeFunnel(app, app.wikiSite, invokeSource)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        disposables.clear()
        _binding = null
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)
        callback()?.onCancelThemeChooser()
    }

    private fun disableButtonsOnMobileWeb() {
        binding.textSizeSeekBar.isEnabled = !isMobileWeb
        binding.buttonDecreaseTextSize.isEnabled = !isMobileWeb
        binding.buttonIncreaseTextSize.isEnabled = !isMobileWeb
        binding.buttonFontFamilySerif.isEnabled = !isMobileWeb
        binding.buttonFontFamilySansSerif.isEnabled = !isMobileWeb
        binding.themeChooserMatchSystemThemeSwitch.isEnabled = !isMobileWeb
        binding.themeChooserDarkModeDimImagesSwitch.isEnabled = !isMobileWeb && binding.themeChooserDarkModeDimImagesSwitch.isEnabled
        binding.themeChooserReadingFocusModeSwitch.isEnabled = !isMobileWeb
        binding.buttonThemeBlack.isEnabled = binding.buttonThemeBlack.isEnabled && (app.currentTheme == Theme.BLACK || !isMobileWeb)
        binding.buttonThemeDark.isEnabled = binding.buttonThemeDark.isEnabled && (app.currentTheme == Theme.DARK || !isMobileWeb)
        binding.buttonThemeLight.isEnabled = app.currentTheme == Theme.LIGHT || !isMobileWeb
        binding.buttonThemeSepia.isEnabled = app.currentTheme == Theme.SEPIA || !isMobileWeb

        if (isMobileWeb) {
            val textColor = ResourceUtil.getThemedColor(requireContext(), R.attr.color_group_61)
            binding.buttonDecreaseTextSize.setTextColor(textColor)
            binding.buttonIncreaseTextSize.setTextColor(textColor)
            binding.buttonFontFamilySerif.setTextColor(textColor)
            binding.buttonFontFamilySerif.setTextColor(textColor)
            binding.themeChooserMatchSystemThemeSwitch.setTextColor(textColor)
            binding.themeChooserDarkModeDimImagesSwitch.setTextColor(textColor)
            binding.themeChooserReadingFocusModeSwitch.setTextColor(textColor)
            binding.themeChooserReadingFocusModeDescription.setTextColor(textColor)
            updateThemeButtonAlpha(binding.buttonThemeBlack, !binding.buttonThemeBlack.isEnabled)
            updateThemeButtonAlpha(binding.buttonThemeDark, !binding.buttonThemeDark.isEnabled)
            updateThemeButtonAlpha(binding.buttonThemeLight, !binding.buttonThemeLight.isEnabled)
            updateThemeButtonAlpha(binding.buttonThemeSepia, !binding.buttonThemeSepia.isEnabled)
        }
    }

    private fun onToggleDimImages(enabled: Boolean) {
        if (enabled == Prefs.dimDarkModeImages) {
            return
        }
        Prefs.dimDarkModeImages = enabled
        callback()?.onToggleDimImages()
    }

    private fun onToggleMatchSystemTheme(enabled: Boolean) {
        if (enabled == Prefs.shouldMatchSystemTheme) {
            return
        }
        Prefs.shouldMatchSystemTheme = enabled
        val currentTheme = app.currentTheme
        if (isMatchingSystemThemeEnabled) {
            when (app.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) {
                Configuration.UI_MODE_NIGHT_YES -> if (!app.currentTheme.isDark) {
                    app.currentTheme = if (!app.unmarshalTheme(Prefs.previousThemeId).isDark) Theme.BLACK else app.unmarshalTheme(Prefs.previousThemeId)
                    Prefs.previousThemeId = currentTheme.marshallingId
                }
                Configuration.UI_MODE_NIGHT_NO -> if (app.currentTheme.isDark) {
                    app.currentTheme = if (app.unmarshalTheme(Prefs.previousThemeId).isDark) Theme.LIGHT else app.unmarshalTheme(Prefs.previousThemeId)
                    Prefs.previousThemeId = currentTheme.marshallingId
                }
            }
        }
        conditionallyDisableThemeButtons()
    }

    private fun onToggleReadingFocusMode(enabled: Boolean) {
        Prefs.readingFocusModeEnabled = enabled
        funnel.logReadingFocusMode(enabled)
        callback()?.onToggleReadingFocusMode()
    }

    private fun conditionallyDisableThemeButtons() {
        updateThemeButtonAlpha(binding.buttonThemeLight, isMatchingSystemThemeEnabled && app.currentTheme.isDark)
        updateThemeButtonAlpha(binding.buttonThemeSepia, isMatchingSystemThemeEnabled && app.currentTheme.isDark)
        updateThemeButtonAlpha(binding.buttonThemeDark, isMatchingSystemThemeEnabled && !app.currentTheme.isDark)
        updateThemeButtonAlpha(binding.buttonThemeBlack, isMatchingSystemThemeEnabled && !app.currentTheme.isDark)
        binding.buttonThemeLight.isEnabled = !isMatchingSystemThemeEnabled || !app.currentTheme.isDark
        binding.buttonThemeSepia.isEnabled = !isMatchingSystemThemeEnabled || !app.currentTheme.isDark
        binding.buttonThemeDark.isEnabled = !isMatchingSystemThemeEnabled || app.currentTheme.isDark
        binding.buttonThemeBlack.isEnabled = !isMatchingSystemThemeEnabled || app.currentTheme.isDark
    }

    private fun updateThemeButtonAlpha(button: View, translucent: Boolean) {
        button.alpha = if (translucent) 0.2f else 1.0f
    }

    private val isMatchingSystemThemeEnabled: Boolean
        get() = Prefs.shouldMatchSystemTheme && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q

    private fun updateComponents() {
        updateFontSize()
        updateFontFamily()
        updateThemeButtons()
        updateDimImagesSwitch()
        updateMatchSystemThemeSwitch()
        disableButtonsOnMobileWeb()

        binding.themeChooserReadingFocusModeSwitch.isChecked = Prefs.readingFocusModeEnabled
    }

    private fun updateMatchSystemThemeSwitch() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            binding.themeChooserMatchSystemThemeSwitch.visibility = View.VISIBLE
            binding.themeChooserMatchSystemThemeSwitch.isChecked = Prefs.shouldMatchSystemTheme
            conditionallyDisableThemeButtons()
        } else {
            binding.themeChooserMatchSystemThemeSwitch.visibility = View.GONE
        }
    }

    private fun updateFontSize() {
        val multiplier = Prefs.textSizeMultiplier
        binding.textSizeSeekBar.value = multiplier
        val percentStr = getString(R.string.text_size_percent,
                (100 * (1 + multiplier * DimenUtil.getFloat(R.dimen.textSizeMultiplierFactor))).toInt())
        binding.textSizePercent.text = if (multiplier == 0) getString(R.string.text_size_percent_default, percentStr) else percentStr
        if (updatingFont) {
            binding.fontChangeProgressBar.visibility = View.VISIBLE
        } else {
            binding.fontChangeProgressBar.visibility = View.GONE
        }
    }

    private fun updateFontFamily() {
        binding.buttonFontFamilySansSerif.strokeWidth = if (Prefs.fontFamily == binding.buttonFontFamilySansSerif.tag) BUTTON_STROKE_WIDTH else 0
        binding.buttonFontFamilySerif.strokeWidth = if (Prefs.fontFamily == binding.buttonFontFamilySerif.tag) BUTTON_STROKE_WIDTH else 0
    }

    private fun updateThemeButtons() {
        updateThemeButtonStroke(binding.buttonThemeLight, app.currentTheme === Theme.LIGHT)
        updateThemeButtonStroke(binding.buttonThemeSepia, app.currentTheme === Theme.SEPIA)
        updateThemeButtonStroke(binding.buttonThemeDark, app.currentTheme === Theme.DARK)
        updateThemeButtonStroke(binding.buttonThemeBlack, app.currentTheme === Theme.BLACK)
    }

    private fun updateThemeButtonStroke(button: MaterialButton, selected: Boolean) {
        button.strokeWidth = if (selected) BUTTON_STROKE_WIDTH else 0
        button.isClickable = !selected
    }

    private fun updateDimImagesSwitch() {
        binding.themeChooserDarkModeDimImagesSwitch.isChecked = Prefs.dimDarkModeImages
        binding.themeChooserDarkModeDimImagesSwitch.isEnabled = app.currentTheme.isDark
        binding.themeChooserDarkModeDimImagesSwitch.setTextColor(if (binding.themeChooserDarkModeDimImagesSwitch.isEnabled)
            ResourceUtil.getThemedColor(requireContext(), R.attr.section_title_color) else ContextCompat.getColor(requireContext(), R.color.black26))
    }

    private inner class ThemeButtonListener(private val theme: Theme) : View.OnClickListener {
        override fun onClick(v: View) {
            if (app.currentTheme !== theme) {
                funnel.logThemeChange(app.currentTheme, theme)
                app.currentTheme = theme
            }
        }
    }

    private inner class FontFamilyListener : View.OnClickListener {
        override fun onClick(v: View) {
            if (v.tag != null) {
                val newFontFamily = v.tag as String
                funnel.logFontThemeChange(Prefs.fontFamily, newFontFamily)
                app.setFontFamily(newFontFamily)
            }
        }
    }

    private inner class FontSizeButtonListener(private val action: FontSizeAction) : View.OnClickListener {
        override fun onClick(view: View) {
            val currentMultiplier = Prefs.textSizeMultiplier
            val changed = when (action) {
                FontSizeAction.INCREASE -> {
                    app.setFontSizeMultiplier(Prefs.textSizeMultiplier + 1)
                }
                FontSizeAction.DECREASE -> {
                    app.setFontSizeMultiplier(Prefs.textSizeMultiplier - 1)
                }
                FontSizeAction.RESET -> {
                    app.setFontSizeMultiplier(0)
                }
            }
            if (changed) {
                updatingFont = true
                updateFontSize()
                funnel.logFontSizeChange(currentMultiplier.toFloat(), Prefs.textSizeMultiplier.toFloat())
            }
        }
    }

    private inner class EventBusConsumer : Consumer<Any> {
        override fun accept(event: Any?) {
            if (event is WebViewInvalidateEvent) {
                updatingFont = false
                updateComponents()
            }
        }
    }

    fun callback(): Callback? {
        return FragmentUtil.getCallback(this, Callback::class.java)
    }

    companion object {
        private const val EXTRA_IS_MOBILE_WEB = "isMobileWeb"
        private val BUTTON_STROKE_WIDTH = DimenUtil.roundedDpToPx(2f)

        fun newInstance(source: InvokeSource, isMobileWeb: Boolean = false): ThemeChooserDialog {
            return ThemeChooserDialog().apply {
                arguments = bundleOf(Constants.INTENT_EXTRA_INVOKE_SOURCE to source,
                    EXTRA_IS_MOBILE_WEB to isMobileWeb)
            }
        }
    }
}
